src/garage/torch/algos/ddpg.py

```diff
@@ -0,0 +1,300 @@
+"""This modules creates a DDPG model in PyTorch."""
+from collections import deque
+
+from dowel import logger, tabular
+import numpy as np
+import torch
+
+from garage.np.algos.off_policy_rl_algorithm import OffPolicyRLAlgorithm
+
+
+class DDPG(OffPolicyRLAlgorithm):
+    """A DDPG model implemented with PyTorch.
+
+    DDPG, also known as Deep Deterministic Policy Gradient, uses actor-critic
+    method to optimize the policy and reward prediction. It uses a supervised
+    method to update the critic network and policy gradient to update the actor
+    network. And there are exploration strategy, replay buffer and target
+    networks involved to stabilize the training process.
+
+    Args:
+        env_spec (EnvSpec): Environment specification.
+        policy (garage.torch.policies.base.Policy): Policy.
+        target_policy (garage.torch.policies.base.Policy): Copy of policy.
+        qf (object): Q-value network.
+        target_qf (object): Copy of Q-value network.
+        replay_buffer (garage.replay_buffer.ReplayBuffer): Replay buffer.
+        n_epoch_cycles (int): Number of train_once calls per epoch.
+        n_train_steps (int): Training steps.
+        max_path_length (int): Maximum path length. The episode will
+            terminate when length of trajectory reaches max_path_length.
+        min_buffer_size (int): The minimum buffer size for replay buffer.
+        rollout_batch_size (int): Roll out batch size.
+        exploration_strategy (garage.np.exploration_strategies.
+            ExplorationStrategy): Exploration strategy.
+        target_update_tau (float): Interpolation parameter for doing the
+            soft target update.
+        policy_lr (float): Learning rate for training policy network.
+        qf_lr (float): Learning rate for training Q-value network.
+        discount(float): Discount factor for the cumulative return.
+        policy_weight_decay (float): L2 weight decay factor for parameters
+            of the policy network.
+        qf_weight_decay (float): L2 weight decay factor for parameters
+            of the q value network.
+        optimizer (torch.optimizer): Optimizer for training policy network
+            and Q-vaule network.
+        criterion (torch.nn.modules.loss): Loss function for Q-value newtork.
+        clip_pos_returns (bool): Whether or not clip positive returns.
+        clip_return (float): Clip return to be in [-clip_return,
+            clip_return].
+        max_action (float): Maximum action magnitude.
+        reward_scale (float): Reward scale.
+        input_include_goal (bool): Whether input includes goal.
+        smooth_return (bool): Whether to smooth the return.
```

Comment:
`Whether to smooth the return for logging.`

This option was introduced just because baseline used this and we want to compare against baseline. We should consider removing this at some point.

Reply:
we will still be benchmarking against baselines for the forseeable future, but if we want to do this a different way (e.g. with inheritance or a preprocessor or a wrapper), that's fine too